pkg/broker/work_engine.go

```diff
@@ -50,10 +50,33 @@ func (engine *WorkEngine) StartNewAsyncJob(
 	if token == "" {
 		token = engine.Token()
 	}
-	go work.Run(token, engine.topic(topic))
+	go engine.start(token, work, topic)
+
 	return token, nil
 }
 
+func (engine *WorkEngine) start(token string, work Work, topic WorkTopic) {
+	engine.jobs[token] = make(chan JobMsg)
+	// run the job and close the channel in a new routine
+
+	defer func() {
+		log.Debug("closing channel for job ", token, engine.jobs)
+		close(engine.jobs[token])
+		delete(engine.jobs, token)
+	}()
+
+	go func() {
+		// listen for new messages and hand them off to the subscribers
+		for msg := range engine.jobs[token] {
+			for _, sub := range engine.subscribers[topic] {
+				//TODO edge case consider the fact that a subscriber may never exit and so we would leak go routines
+				go sub.Notify(msg)
```

Comment:
In regards to the comment, I do think that having a context timeout makes sense here. 

I think a simple implementation would be to wrap this in a func() that waited for a configurable amount of time. 

The notify interface could then have an error return that we would log if something does go wrong. and if it returns with nil then everythng is all good. 

```go
...
for _, sub := range engine.subscribers[topic] {
         c := make(chan error, 1)
        go waitForNotify(engine.duration, sub, msg)
....

func waitForNotify(d time.Duration,  w WorkSubscriber, j JobMessage) {
              select {
            case err := w.Notify(j):
                 handle error or if nil return
            case <-d:
            }
}

```

Reply:
Going to chew this one over 🤔

Reply:
I think your right, I think you need to pass a channel to `waitForNotify` function and add the timeout bits to a function that gets launched with the go routine above.

Something like this?
https://play.golang.org/p/pRjyCXfZgSi

Reply:
updated to take into account your feedback plus look at adding in the wait group and using context https://gist.github.com/maleck13/d2c903eac6064422966d64715118345e